### ISSUE TITLE
**故障注入调研与实验：王星艺**

### ISSUE THREAD

**Issue by ACA-Lab-SJTU** [acada-sjtu/SysReliability2019#5]
-[] 完成openRISC及Leon的故障注入实验
-[] 完成ITC paper相关部分的撰写
-[] 协助李渝进行其他故障注入方法研究

**Comment by JarvisXX** [comment]
Try to parse OpenRISC and Leon3. 
Since yosys tools didn't work out in converting Verilog to BENCH, I started to write scripts to do such conversion. The netlist file is well-organized, and I've parsed the inputs, outputs. However, the major obstacle is the logic functions. I think the netlists that Prof. Krish provides are pre-treated. The basic logic functions, such as AND, OR, and INV, are easy to understand. But some of them are not quite clear to me. I'll try to understand the meaning of the unknown operators before ask Prof. Krish for help.

**Comment by JarvisXX** [comment]
![OR1200](https://user-images.githubusercontent.com/11536724/54992986-c0b53f00-4ffb-11e9-9e4c-511ce8e545e0.png)
![LEON3](https://user-images.githubusercontent.com/11536724/54992989-c27f0280-4ffb-11e9-87fc-e6076caf9208.png)

**Comment by JarvisXX** [comment]
The converting script is complete. Now I'm generating the Dataset on OpenRISC (or1200). The speed is about 3min/sample. I use multi-process programming to generate them, and by tonight, I can get about 2000 samples. I'll try to training some models based on these samples to see the performance.

**Comment by ACA-Lab-SJTU** [comment]
Good job. 

-
Li Jiang, Ph.D., Associate Professor,
Department of computer science and engineering,
Shanghai Jiao Tong University

**Comment by JarvisXX** [comment]
Experimental results on OR1200 (partly).
OR1200 has 338 inputs, 217 outputs, 49018 gates and 3073 FFs. This is very large compared to ITC99 whose largest circuit is only about 8000 gates and 200 FFs. 

I did the experiments on OR1200 with 4200 samples. The training and testing samples are 80%-20%, and the faulty and fault-free ones are 50%-50%. 

At first, I input all the 3073 FFs into LSTM model, and the accuracy is about 50% (randomly guessing). Then I tried to do the FF selection. The results are as follows:
- number of FFs selected: 1000, accuracy: 55%
- number of FFs selected: 500, accuracy: 56%
- number of FFs selected: 200, accuracy: 54%
It seems that 500 is a good option. But in the training process, both training and validation accuracy is jumping between 50% and 55%. I think the reason behind this is still the lack of training samples. About 3000 samples are not enough for LSTM to converge. 

Another phenomenon is that, the precision and recall is not stable, too. With the accuracy being 55%, the precision and recall can be 97% & 10%, and 58% & 41% in another case.

I also did BiLSTM experiments, and the results are almost the same.

In conclusion, the number of training samples is the major problem. The fault injection process is still running. I'll continue to update the experimental results.

**Comment by JarvisXX** [comment]
About the paper:
I've revised the paper on Overleaf according to Prof. Krish's comments. However, some questions are hard for me to answer. I've marked them as comments on Overleaf, too.
